env/env_test.cc

```diff
@@ -579,142 +593,127 @@ TEST_P(EnvPosixTestWithParam, DecreaseNumBgThreads) {
+  SleepUntilQueueLen(Env::Priority::HIGH, 0U);
 
   // Enqueue 5 more tasks. Thread pool size now is 4.
   // Task 0, 3, 4, 5 running;6, 7 waiting.
   for (size_t i = 3; i < 8; i++) {
     env_->Schedule(&test::SleepingBackgroundTask::DoSleepTask, &tasks[i],
                    Env::Priority::HIGH);
   }
-  Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_EQ(2U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
-  ASSERT_TRUE(tasks[3].IsSleeping());
-  ASSERT_TRUE(tasks[4].IsSleeping());
-  ASSERT_TRUE(tasks[5].IsSleeping());
-  ASSERT_TRUE(!tasks[6].IsSleeping());
-  ASSERT_TRUE(!tasks[7].IsSleeping());
+  SleepUntilQueueLen(Env::Priority::HIGH, 2U);
+  ASSERT_TRUE(tasks[3].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[4].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[5].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[6].ShortWaitForWokenUp());
+  ASSERT_TRUE(tasks[7].ShortWaitForWokenUp());
 
   // Wake up task 0, 3 and 4. Task 5, 6, 7 running.
   tasks[0].WakeUp();
   tasks[3].WakeUp();
   tasks[4].WakeUp();
 
-  Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_EQ((unsigned int)0, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+  SleepUntilQueueLen(Env::Priority::HIGH, 0U);
   for (size_t i = 5; i < 8; i++) {
-    ASSERT_TRUE(tasks[i].IsSleeping());
+    ASSERT_TRUE(tasks[i].ShortWaitForSleeping());
   }
 
   // Shrink back to 1 thread. Still task 5, 6, 7 running
   env_->SetBackgroundThreads(1, Env::Priority::HIGH);
   Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_TRUE(tasks[5].IsSleeping());
-  ASSERT_TRUE(tasks[6].IsSleeping());
-  ASSERT_TRUE(tasks[7].IsSleeping());
+  ASSERT_TRUE(tasks[5].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[6].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[7].ShortWaitForSleeping());
 
   // Wake up task  6. Task 5, 7 running
   tasks[6].WakeUp();
   Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_TRUE(tasks[5].IsSleeping());
-  ASSERT_TRUE(!tasks[6].IsSleeping());
-  ASSERT_TRUE(tasks[7].IsSleeping());
+  ASSERT_TRUE(tasks[5].ShortWaitForSleeping());
+  ASSERT_TRUE(tasks[6].ShortWaitForWokenUp());
+  ASSERT_TRUE(tasks[7].ShortWaitForSleeping());
 
   // Wake up threads 7. Task 5 running
   tasks[7].WakeUp();
   Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_TRUE(!tasks[7].IsSleeping());
+  ASSERT_TRUE(tasks[7].ShortWaitForWokenUp());
 
   // Enqueue thread 8 and 9. Task 5 running; one of 8, 9 might be running.
   env_->Schedule(&test::SleepingBackgroundTask::DoSleepTask, &tasks[8],
                  Env::Priority::HIGH);
   env_->Schedule(&test::SleepingBackgroundTask::DoSleepTask, &tasks[9],
                  Env::Priority::HIGH);
-  Env::Default()->SleepForMicroseconds(kDelayMicros);
-  ASSERT_GT(env_->GetThreadPoolQueueLen(Env::Priority::HIGH), (unsigned int)0);
-  ASSERT_TRUE(!tasks[8].IsSleeping() || !tasks[9].IsSleeping());
+  ASSERT_TRUE(tasks[8].ShortWaitForWokenUp() || tasks[9].ShortWaitForWokenUp());
```

Comment:
At this point, thread pool size is 1 and task 5 is running (sleep_ is true). I think should assert the following
```
ASSERT_TRUE(!tasks[8].IsSleeping());
ASSERT_TRUE(!tasks[9].IsSleeping());
```